src/shims/fs.rs

```diff
@@ -722,6 +799,266 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                //     pub d_off: off64_t,
+                //     pub d_reclen: c_ushort,
+                //     pub d_type: c_uchar,
+                //     pub d_name: [c_char; 256],
+                // }
+
+                let entry_place = this.deref_operand(entry_op)?;
+                let name_place = this.mplace_field(entry_place, 4)?;
+
+                let file_name = dir_entry.file_name();
+                let (name_fits, _) = this.write_os_str_to_c_str(
+                    &file_name, name_place.ptr,
+                    name_place.layout.size.bytes(),
+                )?;
+                if !name_fits {
+                    throw_unsup_format!("A directory entry had a name too large to fit in libc::dirent64");
+                }
+
+                let entry_place = this.deref_operand(entry_op)?;
+                let ino64_t_layout = this.libc_ty_layout("ino64_t")?;
+                let off64_t_layout = this.libc_ty_layout("off64_t")?;
+                let c_ushort_layout = this.libc_ty_layout("c_ushort")?;
+                let c_uchar_layout = this.libc_ty_layout("c_uchar")?;
+
+                // If the host is a Unix system, fill in the inode number with its real value.
+                // If not, use 0 as a fallback value.
+                #[cfg(unix)]
+                let ino = std::os::unix::fs::DirEntryExt::ino(&dir_entry);
+                #[cfg(not(unix))]
+                let ino = 0;
+
+                let file_type = this.file_type_to_d_type(dir_entry.file_type())? as u128;
+
+                let imms = [
+                    immty_from_uint_checked(ino, ino64_t_layout)?, // d_ino
+                    immty_from_uint_checked(0u128, off64_t_layout)?, // d_off
+                    immty_from_uint_checked(0u128, c_ushort_layout)?, // d_reclen
+                    immty_from_uint_checked(file_type, c_uchar_layout)?, // d_type
+                ];
+                this.write_packed_immediates(entry_place, &imms)?;
+
+                let result_place = this.deref_operand(result_op)?;
+                this.write_scalar(this.read_scalar(entry_op)?, result_place.into())?;
+
+                Ok(0)
+            }
+            None => {
+                // end of stream: return 0, assign *result=NULL
+                this.write_null(this.deref_operand(result_op)?.into())?;
+                Ok(0)
+            }
+            Some(Err(e)) => match e.raw_os_error() {
+                // return positive error number on error
+                Some(error) => Ok(error),
+                None => {
+                    throw_unsup_format!("The error {} couldn't be converted to a return value", e)
+                }
+            },
+        }
+    }
+
+    fn macos_readdir_r(
+        &mut self,
+        dirp_op: OpTy<'tcx, Tag>,
+        entry_op: OpTy<'tcx, Tag>,
+        result_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("readdir_r")?;
```

Comment:
We now also have a utility function double-checking that this is called only on macos, `this.assert_platform("macos", "readdir_r");`, could you call that here? (And same for the `linux_` function.)